### PR TITLE
[NNC] Make splitWithTail transform in-place

### PR DESCRIPTION
**Pull request by navahgar**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #58270 [NNC] Add documentation for splitWith APIs
* #58269 [NNC] Make splitWithMask transform in-place
* **#58268 [NNC] Make splitWithTail transform in-place**

Differential Revision: [D28427228](https://our.internmc.facebook.com/intern/diff/D28427228)

### CODE REVIEW

**Comment by navahgar** [comment]
@navahgar has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28427228).

**Comment by navahgar** [comment]
@navahgar has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28427228).

File: test/cpp/tensorexpr/test_loopnest.cpp [hunk 1]
```diff
@@ -698,7 +703,7 @@ TEST(LoopNest, SplitWithTailWithLoopOptions) {
   ASSERT_GT(loops.size(), 0);
   l.setGPUBlockIndex(loops[0], LoopOptions::IDX_Y);
   l.splitWithTail(loops[0], 4, &outer, &inner, &tail);
-  ASSERT_NE(outer, nullptr);
+  ASSERT_EQ(loops[0], outer);
```
Comment (ZolotukhinM) [inline test/cpp/tensorexpr/test_loopnest.cpp:706]: Please also add a test checking that loops from the body remain valid after the split as well (see a comment to the second PR in the stack for an example).

**Comment by navahgar** [comment]
@navahgar has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28427228).

File: test/cpp/tensorexpr/test_loopnest.cpp (hunk 1 above)
Comment (navahgar) [reply test/cpp/tensorexpr/test_loopnest.cpp]: Will add this test once `IRMutator` and `Substitute` APIs are fixed.

File: torch/csrc/jit/tensorexpr/tensorexpr_init.cpp [hunk 2]
```diff
@@ -413,9 +413,9 @@ void initTensorExprBindings(PyObject* module) {
       .def(
           "split_with_tail",
           [](const LoopNest& self, For* f, int factor) {
-            For *outer = nullptr, *inner = nullptr, *tail = nullptr;
-            self.splitWithTail(f, factor, &outer, &inner, &tail);
-            return std::make_tuple(outer, inner, tail);
+            For *inner = nullptr, *tail = nullptr;
+            self.splitWithTail(f, factor, &inner, &tail);
+            return std::make_tuple(inner, tail);
```
Comment (navahgar) [inline torch/csrc/jit/tensorexpr/tensorexpr_init.cpp:418]: @ZolotukhinM Does this change require modifications anywhere else?

Comment (navahgar) [reply torch/csrc/jit/tensorexpr/tensorexpr_init.cpp:418]: Yeah. I checked those already. They all pass.

File: benchmarks/cpp/tensorexpr/bench_reduce.cpp [hunk 3]
```diff
@@ -266,10 +266,9 @@ BENCHMARK_DEFINE_F(Reduce1D, TeSplitTail)(benchmark::State& state) {
   {
     auto const& loops = loop.getLoopStmtsFor(BT);
     te::For* m = loops[1];
-    te::For* mo;
     te::For* mi;
     te::For* tail;
-    loop.splitWithTail(m, kChunkSize, &mo, &mi, &tail);
+    loop.splitWithTail(m, kChunkSize, &mi, &tail);
```
Comment (ZolotukhinM) [inline benchmarks/cpp/tensorexpr/bench_reduce.cpp:271]: Nit: can remove `mi` and `tail` entirely since they are unused (here and in other places too).

**Review by ZolotukhinM — APPROVED** [review]
👍

**Comment by navahgar** [comment]
@navahgar has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28427228).

File: benchmarks/cpp/tensorexpr/bench_reduce.cpp (hunk 3 above)
Comment (navahgar) [reply benchmarks/cpp/tensorexpr/bench_reduce.cpp]: Done.

**Comment by navahgar** [comment]
@navahgar has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28427228).

**Pull request closed without merging**